### PR TITLE
Riemann worker

### PR DESCRIPTION
**Pull request by bazsi**
This branch reworks the riemann module so that it does not use logthrdestdrv's compatibility mode, e.g.
it contains the split between RiemannDestDriver and RiemannDestWorker.

At the same time it adds support for flush-timeout().

I performed a few simple manual tests to confirm that the functionality still works, but I am not using riemann in production.

### CODE REVIEW

**Comment by bazsi** [comment]
NOTE: this depends on http-multi-threading, which should be merged firs.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by furiel** [comment]
@kira-syslogng retest this please

File: modules/riemann/Makefile.am [hunk 1]
```diff
@@ -12,6 +12,8 @@ modules_riemann_libriemann_la_SOURCES		 = \
 	modules/riemann/riemann-grammar.y	   \
 	modules/riemann/riemann.c		   \
 	modules/riemann/riemann.h		   \
+	modules/riemann/riemann-worker.c	   \
```
Comment (furiel) [inline modules/riemann/Makefile.am]: Can you add these to CMakeLists.txt as well?

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by bazsi** [comment]
@kira-syslogng retest this please;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by furiel** [comment]
@bazsi it seems we leak the a GCond somewhere:

```
@version: 3.18

options {
    threaded(yes);
};

destination d_riemann {
    riemann(description($MESSAGE)
    port(5555)
    server("localhost"));
};

log {
  source { stdin(flags(no-parse)); };
  destination(d_riemann);

  flags(flow-control);
};
```

```
==14143== 16 bytes in 1 blocks are definitely lost in loss record 21 of 66
==14143==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14143==    by 0x51847B8: g_malloc (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4800.2)
==14143==    by 0x519B9C2: g_slice_alloc (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4800.2)
==14143==    by 0x515179A: g_cond_new (in /lib/x86_64-linux-gnu/libglib-2.0.so.0.4800.2)
==14143==    by 0x4E87276: log_threaded_dest_worker_init_instance (logthrdestdrv.c:696)
==14143==    by 0x778A986: riemann_dw_new (riemann-worker.c:427)
==14143==    by 0x4E875A4: _construct_worker (logthrdestdrv.c:799)
==14143==    by 0x4E876AD: _start_worker_thread (logthrdestdrv.c:819)
==14143==    by 0x4E87CD5: log_threaded_dest_driver_start_workers (logthrdestdrv.c:968)
==14143==    by 0x77893A8: riemann_dd_init (riemann.c:289)
==14143==    by 0x4E73C02: log_pipe_init (logpipe.h:292)
==14143==    by 0x4E761D5: cfg_tree_start (cfg-tree.c:1392)
```

**Comment by bazsi** [comment]
@furiel a chaining call to the inherited destructor was missing. I've added that to the latest round, thanks for catching it.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/riemann/riemann-grammar.ym [hunk 2]
```diff
@@ -140,6 +140,10 @@ riemann_option
           {
             log_threaded_dest_driver_set_flush_lines(last_driver,  $3);
           }
+        | KW_FLUSH_TIMEOUT '(' nonnegative_integer ')'
```
Comment (Kokan) [inline modules/riemann/riemann-grammar.ym]: Would it be possible to move this to the `lib/cfg-grammar.y` something like options for threaded dest drivers ?

Comment (gaborznagy) [reply modules/riemann/riemann-grammar.ym]: IMHO This is resolved by #2386

Comment (bazsi) [reply modules/riemann/riemann-grammar.ym]: I have now rebased this branch, and indeed flush-*() related options are now handled by a rule
in cfg-grammar, which is just referenced from here.

This resolves this issue.

File: modules/riemann/riemann.c [hunk 3]
```diff
@@ -699,9 +509,200 @@ riemann_dd_init(LogPipe *s)
...
+
+  self->port = port;
+}
+
+void
+riemann_dd_set_field_host(LogDriver *d, LogTemplate *value)
+{
+  RiemannDestDriver *self = (RiemannDestDriver *)d;
+
+  self->fields.host = log_template_ref(value);
```
[the remark this answers is not in the document]
Comment (bazsi) [reply modules/riemann/riemann.c]: that leak has been there even before the refactor, but I've now fixed it with an extra patch on top.

**Comment by bazsi** [comment]
I have fixed all outstanding notes and rebased it against master.

**Comment by bazsi** [comment]
This should be good enough for merging, @furiel approved this originally, which has become stale because I fixed @Kokan's review comments, but that should not be a material change.

So if both @Kokan and @furiel approves, this could go in. (which I'd appreciate as the batch-* related PR caused ugly-to-resolve conflicts, again).

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Pull request merged**